### PR TITLE
📇 GRPO: print completions to console and update docs

### LINKED ISSUE #2948: GPRO: Expand log_completions logic or update docs
### Feature request

Currently, the GRPOConfig docstring for `log_completions` is:
> log_completions (bool, optional, defaults to False) — Whether to log the completions during training.

Which is a bit misleading, given that these are only logged when Weights and Biases is active.

Suggestion 1: Change the docs:
> log_completions (bool, optional, defaults to False) — If True, will try to log generated (prompt, completion) pairs to `wandb`, if it's available.

Suggestion 2: Log to `wandb` if possible, but also log to console. Since `rich` is a requirement, it could be a rich table:

```python
from rich.console import Console
from rich.panel import Panel
from rich.table import Table
from rich.text import Text

def print_output_sample(prompts: list[str], completions: list[str]):
    """Print out a sample of model completions."""
    console = Console()
    table = Table(show_header=True, header_style="bold white", expand=True, padding=(0, 1, 1, 0))
    table.add_column("Prompt", style="bright_yellow")
    for s, p in zip(source, prediction, strict=True):
        table.add_row(Text(s), Text(p))
    panel = Panel(table, expand=False, title="Output Sample", border_style="bold white")
    console.print(panel)
```

In that case, it may be useful to let `log_completions` be an int to specify a custom interval and avoid the console being spammed.

### Motivation

It's only right

### Your contribution

See suggestions

### PR DESCRIPTION
**Pull request by nopepper**
# What does this PR do?

- Update `GRPOConfig` to replace `log_completions` with `log_completions_steps`
- Add `print_prompt_completions_sample()` utility function for rich console logging
- Modify `GRPOTrainer` to additionally print 5 random prompt-completion pairs every log_completions_steps steps

Fixes #2948 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -839,22 +846,34 @@ def _generate_and_score_completions(
         self._metrics[mode]["reward_std"].append(std_grouped_rewards.mean().item())
 
         if (
-            self.log_completions
-            and self.state.global_step % self.args.logging_steps == 0
-            and "wandb" in self.args.report_to
+            self.log_completions_steps is not None
+            and self.log_completions_steps > 0
+            and self.state.global_step % self.log_completions_steps == 0
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:851]: why do you replace this

```diff
- self.state.global_step % self.args.logging_steps == 0
+ self.state.global_step % self.log_completions_steps == 0
```

logging steps seems sufficient here

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -839,22 +846,34 @@ def _generate_and_score_completions(
...
+            print_prompt_completions_sample(
+                [prompt for prompt, _ in pairs_to_log],
+                [completion for _, completion in pairs_to_log],
+                self.state.global_step,
+            )
+            if (
+                self.args.report_to
+                and "wandb" in self.args.report_to
+                and wandb.run is not None
+                and self.accelerator.is_main_process
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:865]: I think this condition should be moved up

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (nopepper) [reply trl/trainer/grpo_trainer.py:851]: The idea was to decouple `log_completions_steps` from `logging_steps` since the latter could lead to a lot prints to the console. The user may not want to see a big table of text every 10 seconds.

We could bind it to `logging_steps` again if you think it's not necessary

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (nopepper) [reply trl/trainer/grpo_trainer.py:865]: Up where? 😅

File: trl/trainer/utils.py [hunk 3]
```diff
@@ -32,7 +32,9 @@
 from accelerate.state import AcceleratorState
 from huggingface_hub import ModelCard, ModelCardData
 from rich.console import Console
+from rich.panel import Panel
 from rich.table import Table
+from rich.text import Text
```
Comment (qgallouedec) [inline trl/trainer/utils.py:37]: It's a bit out of the scope, but I'd like to make rich an optional dep of trl. Can you use this instead:

```python
from ..import_utils import is_rich_available

if is_rich_available():
    from rich.console import Console
    from rich.panel import Panel
    from rich.table import Table
    from rich.text import Text

...

def print_prompt_completions_sample(...):
    if not is_rich_available():
        raise ImportError("...")
```

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:851]: Yes I think it's not necessary, `logging_step` seems necessary. We could still decouple them later if it's raelly to much noise in the log

File: trl/trainer/utils.py (hunk 3 above)
Comment (nopepper) [reply trl/trainer/utils.py]: I'll do you one better and add a fallback using regular print `print` :-)

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py:865]:
```python
if self.accelerator.is_main_process:
    print_prompt_completions_sample(
        [prompt for prompt, _ in pairs_to_log],
        [completion for _, completion in pairs_to_log],
        self.state.global_step,
    )
    if (
    self.args.report_to
    and "wandb" in self.args.report_to
    and wandb.run is not None
    ):
        ...

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -839,22 +846,34 @@ def _generate_and_score_completions(
...
-                "prompt": gather_object(prompts_text),
-                "completion": gather_object(completions_text),
-                "reward": rewards.tolist(),
-            }
-            df = pd.DataFrame(table)
-
-            if wandb.run is not None and self.accelerator.is_main_process:
+            prompts_to_log: list[str] = gather_object(prompts_text)
+            completions_to_log: list[str] = gather_object(completions_text)
+            pairs_to_log = random.sample(list(zip(prompts_to_log, completions_to_log)), 5)
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:855]: similarly, I'd go for full logging. We can still add a param to limit the number of lines to be logged later.

File: trl/trainer/utils.py [hunk 5]
```diff
@@ -1683,3 +1688,38 @@ def selective_log_softmax(logits, index):
...
+    If `rich` is not available, uses the regular print function.
+
+    Args:
+        prompts (list[str]): List of input prompts.
+        completions (list[str]): List of model-generated completions corresponding to the prompts.
+        step (int): Current training step number, used in the output title.
+
+    Returns:
+        None: Prints the formatted table to the console.
+    """
```
Comment (qgallouedec) [inline trl/trainer/utils.py:1708]:
```suggestion
    """
    Print out a sample of model completions to the console.

    This function creates a nicely formatted table showing prompt-completion pairs, useful for monitoring model outputs
    during training. It requires the `rich` library to be installed.

    Args:
        prompts (`list[str]`):
            List of prompts.
        completions (`list[str]`):
            List of completions corresponding to the prompts.
        step (`int`):
            Current training step number, used in the output title.

    Example:
    ```python
    >>> from trl.trainer.utils import print_prompt_completions_sample
    >>> prompts = ["The sky is", "The sun is"]
    >>> completions = [" blue.", " in the sky."]
    >>> print_prompt_completions_sample(prompts, completions, step=42)
    ╭────────── Step 42 ──────────╮
    │ ┏━━━━━━━━━━━┳━━━━━━━━━━━━━┓ │
    │ ┃Prompt     ┃Completion   ┃ │
    │ ┃           ┃             ┃ │
    │ ┡━━━━━━━━━━━╇━━━━━━━━━━━━━┩ │
    │ │The sky is │ blue.       │ │
    │ │           │             │ │
    │ │The sun is │ in the sky. │ │
    │ │           │             │ │
    │ └───────────┴─────────────┘ │
    ╰─────────────────────────────╯
    ```
    """
```

File: trl/trainer/utils.py (hunk 3 above)
Comment (nopepper) [reply trl/trainer/utils.py]: Ah, it seems like the pre-commit command has issues with the `print` statements. Can we replace them with something else?

File: trl/trainer/utils.py [hunk 6]
```diff
@@ -1683,3 +1688,38 @@ def selective_log_softmax(logits, index):
...
+    if is_rich_available():
+        console = Console()
+        table = Table(show_header=True, header_style="bold white", expand=True, padding=(0, 1, 1, 0))
+        table.add_column("Prompt", style="bright_yellow")
+        table.add_column("Completion", style="bright_green")
+        for s, p in zip(prompts, completions, strict=True):
+            table.add_row(Text(s), Text(p))
```
Comment (qgallouedec) [inline trl/trainer/utils.py:1715]:
```suggestion
        for prompt, completion in zip(prompts, completions, strict=True):
            table.add_row(Text(prompt), Text(completion))
```

File: trl/trainer/utils.py (hunk 3 above)
Comment (nopepper) [reply trl/trainer/utils.py]: Nevermind, I removed it in the end

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (nopepper) [reply trl/trainer/grpo_trainer.py]: Oh true, I missed that. Thanks.

File: trl/trainer/grpo_trainer.py (hunk 1 above)
Comment (nopepper) [reply trl/trainer/grpo_trainer.py]: Alright, I reversed it. Now it should also be compatible with the old API.

File: trl/trainer/grpo_config.py [hunk 7]
```diff
@@ -112,9 +112,8 @@ class GRPOConfig(TrainingArguments):
             set `sync_ref_model=True`.
 
         > Parameters that control the logging
-
         log_completions (`bool`, *optional*, defaults to `False`):
-            Whether to log the completions during training.
+            Print a sample of (prompt, completion) pairs every `logging_steps` steps. Also logs to `wandb` if available.
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py:116]:
```suggestion
        > Parameters that control the logging

        log_completions (`bool`, *optional*, defaults to `False`):
            Whether to log a sample of (prompt, completion) pairs every `logging_steps` steps. If `rich` is installed,
            it prints the sample. If `wandb` logging is enabled, it logs it to `wandb`.
```

File: trl/trainer/grpo_config.py [hunk 8]
```diff
@@ -269,5 +268,7 @@ class GRPOConfig(TrainingArguments):
     # Parameters that control the logging
     log_completions: bool = field(
         default=False,
-        metadata={"help": "Whether to log the completions during training."},
+        metadata={
+            "help": "Print a sample of (prompt, completion) pairs every `logging_steps` steps. Also logs to `wandb` if available."
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py:272]:
```suggestion
            "help": "Whether to log a sample of (prompt, completion) pairs every `logging_steps` steps. If `rich` is "
            "installed, it prints the sample. If `wandb` logging is enabled, it logs it to `wandb`."
```

**Review by qgallouedec — CHANGES_REQUESTED** [review]

**Comment by qgallouedec** [comment]
LGTM, once the latest nit recommendations are applied, and CI green, we're good to merge, thanks :)

File: trl/trainer/grpo_config.py (hunk 7 above)
Comment (nopepper) [reply trl/trainer/grpo_config.py]: I had to change this manually, GitHub doesn't allow committing the suggestion for some reason.

**Comment by qgallouedec** [comment]
Screenshot 2025-02-24 at 23 12 05

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
@bot /style

**Comment by qgallouedec** [comment]
Failing tests are because liger-kernel introduced a bug in their latest version. We can safely ignore it, I guess they'll add a fix soon. See https://github.com/linkedin/Liger-Kernel/issues/586
